### DISCUSSION TITLE
Running PTX backend on WSL 2

### DISCUSSION THREAD

**Discussion by andrii0lomakin** [beehive-lab/TornadoVM#530] [General]
Good day.
I have read in other discussions that there is no guide at the moment on how to run TornadoVM PTX backed on WSL 2. 
So I want to share my experience.
First of all proof that TornadoVM PTX really works on WSL 2.
```
andrii:tornado --devices
WARNING: Using incubator modules: jdk.incubator.vector

Number of Tornado drivers: 1
Driver: PTX
  Total number of PTX devices  : 2
  Tornado device=0:0  (DEFAULT)
        PTX -- PTX -- NVIDIA GeForce RTX 4090
                Global Memory Size: 24.0 GB
                Local Memory Size: 48.0 KB
                Workgroup Dimensions: 3
                Total Number of Block Threads: [2147483647, 65535, 65535]
                Max WorkGroup Configuration: [1024, 1024, 64]
                Device OpenCL C version: N/A

  Tornado device=0:1
        PTX -- PTX -- NVIDIA GeForce RTX 3050 Ti Laptop GPU
                Global Memory Size: 4.0 GB
                Local Memory Size: 48.0 KB
                Workgroup Dimensions: 3
                Total Number of Block Threads: [2147483647, 65535, 65535]
                Max WorkGroup Configuration: [1024, 1024, 64]
                Device OpenCL C version: N/A
```                

Below are steps to achieve that result:
1. Download the nVidia driver from the website https://www.nvidia.com/Download/index.asp .
2. Install the CUDA toolkit from web site https://developer.nvidia.com/cuda-downloads by choosing Linux->WSL 2 from the list of provided OSs. 
3. Add the following script in `~/.bashrc`
```
if [ -d $HOME/.local/bin ]; then
        export PATH=$HOME/.local/bin:$PATH
fi

export CUDA_HOME=/usr/local/cuda

if [ -d $CUDA_HOME/bin ]; then
        export PATH=$CUDA_HOME/bin:$PATH
fi

if [ -d $CUDA_HOME/lib64 ]; then
        export LD_LIBRARY_PATH=$CUDA_HOME/lib64:$LD_LIBRARY_PATH
fi
```
5. Run TornadoVM installer as described in READ.ME .

P.S. That allows me to test my contribution for all backends in one environment. Also, TornadoVM scripts are oriented on a Linux environment to develop there. I hope this helps other contributors.

**Comment by jjfumero** [comment]
That's great. Thanks. I will include this into the `docs`.

**Reply by andrii0lomakin** [reply]
If you use IntelliJ for development, I suggest using at leat 2022.04  it runs on JBR 21 that includes support for Wayland (which is default in WSL 2)
